### PR TITLE
[SD-1170] Mounting algebra

### PR DESCRIPTION
**Pull request by wemrysi**
Adds a `Mounting` algebra describing the operations of managing filesystem mounts. A spec is included, implemented in terms of the algebra that describes how an interpreter of the algebra must behave.

A single interpreter is also defined, that maintains mount state in memory, allowing effectful mounting via a callback function.

### CODE REVIEW

**Comment by wemrysi** [comment]
@sellout @mossprescott @jedesah @drostron This is the direction I'm currently taking with the mounting algebra, if you have a few minutes I'd appreciate any comments, reservations, etc.

File: core/src/main/scala/quasar/mount/Mounting.scala [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+sealed trait Mounting[A]
+
+object Mounting {
+  final case class Lookup(path: APath)
+    extends Mounting[Option[MountConfig2]]
+
+  final case class MountView(loc: AFile, query: Expr)
+    extends Mounting[MountingError \/ Unit]
+
+  final case class MountFileSystem(loc: ADir, typ: FileSystemType, config: Json)
```
Comment (mossprescott) [inline core/src/main/scala/quasar/mount/Mounting.scala]: Given what we know, it's probably safe to make this take just the type and a `connectionUri`.

File: core/src/main/scala/quasar/mount/Mounting.scala [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import scalaz.syntax.apply._
+import scalaz.syntax.monadError._
+
+sealed trait Mounting[A]
+
+object Mounting {
+  final case class Lookup(path: APath)
+    extends Mounting[Option[MountConfig2]]
+
+  final case class MountView(loc: AFile, query: Expr)
```
Comment (mossprescott) [inline core/src/main/scala/quasar/mount/Mounting.scala]: Just a heads up that this becomes `query: Expr, variables: Variables` in my PR coming up probably today.

File: core/src/main/scala/quasar/fp/prism.scala [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package quasar
```
Comment (mossprescott) [inline core/src/main/scala/quasar/fp/prism.scala]: By the way, why are these new files not getting the Apache license head comment? Did we break that plugin?

File: core/src/main/scala/quasar/fp/prism.scala [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+package fp
```
Comment (mossprescott) [inline core/src/main/scala/quasar/fp/prism.scala]: We haven'y used these "exploded" package declarations, as far a I know. I'd say let's be consistent and use the on-liner unless there's a reason to switch.

File: core/src/main/scala/quasar/fp/prism.scala (hunk 3 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/fp/prism.scala]: Hm, it would seem so.

File: core/src/main/scala/quasar/fp/prism.scala (hunk 4 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/fp/prism.scala]: Less explicit importing, afaik.

Comment (jdegoes) [inline core/src/main/scala/quasar/fp/prism.scala]: > Less explicit importing, afaik.

Anti-feature. :smiling_imp:

Comment (sellout) [inline core/src/main/scala/quasar/fp/prism.scala]: Yeah, I agree with @jdegoes. Especially in this case, we don’t want anything else to leak into FP. It should be at the top of the dependency graph.

**Comment by mossprescott** [comment]
Will there be a way to enumerate mounts in this algebra? Because they will certainly show up in QueryFile's `ls`, which makes these two algebras oddly asymmetrical views on the same object. I wonder if they should be more closely integrated.

@jdegoes has talked about a "metadata store" (see [SD-979](https://slamdata.atlassian.net/browse/SD-979)), which would seem to encompass both.

File: core/src/main/scala/quasar/fp/prism.scala (hunk 4 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/fp/prism.scala]: In general, I agree, though for a type in `a.b.c` having members of `a` and `b` in scope seems reasonable given the hierarchical structure. However, I don't hold this strongly so am fine with changing it.

Comment (mossprescott) [inline core/src/main/scala/quasar/fp/prism.scala]: Ah, I see. Amazing how there's a new syntax trap to learn about every day in Scala. :-P

Comment (sellout) [inline core/src/main/scala/quasar/fp/prism.scala]: I think the useful imports tends to go the other way … a package is most likely to use its subpackages, and any sibling/parent package imports need to be more explicit. Ideally, siblings would be independent, and parent packages would contain the parts that tie them together.

**Comment by wemrysi** [comment]
@mossprescott There could be a way to enumerate mounts, though not sure that is needed by anything at the moment. Though you raise an interesting point as to whether `ls` should be concerned with mounts, or just list paths, especially since the filesystem algebras have no notion of mounting.

File: core/src/main/scala/quasar/fp/prism.scala (hunk 4 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/fp/prism.scala]: That's true, though you also tend to see abstractions further up in the hierarchy and then implementations of said abstractions further down.

File: core/src/main/scala/quasar/mount/Mounting.scala (hunk 1 above)
Comment (wemrysi) [inline core/src/main/scala/quasar/mount/Mounting.scala]: Sounds reasonable, if that's the case, then what about changing the JSON representation as well:

```json
{
  "/some/path": { "mongodb": "mongodb://..." },
  "/another/path": { "view": "sql2:///..." }
}
```

i.e., do we still need the intermediate object with the `connectionUri` key?

Comment (wemrysi) [inline core/src/main/scala/quasar/mount/Mounting.scala]: Or even

```json
{
  "/some/path": ["mongodb", "mongodb://..."],
  "/another/path": ["view", "sql2:///..."]
}
```

Comment (mossprescott) [inline core/src/main/scala/quasar/mount/Mounting.scala]: Sure, except that it's a breaking change for the front-end to deal with, as well as making every existing user's configuration invalid, so I guess I don't see the point unless we're making some significant improvement, which I don't think this is.

Comment (wemrysi) [inline core/src/main/scala/quasar/mount/Mounting.scala]: True, and the existing one is more forward-compatible.

**Comment by wemrysi** [comment]
Looking into that compile failure...

**Comment by drostron** [comment]
👍🏽

**Pull request merged**